source/particles/generators.cc

```diff
@@ -75,6 +142,263 @@ namespace Particles
+                                   accumulated_cell_weights.end(),
+                                   random_weight);
+                const unsigned int cell_index =
+                  std::distance(accumulated_cell_weights.begin(),
+                                selected_cell);
+
+                ++particles_per_cell[cell_index];
+              }
+          }
+        else
+          {
+            // Compute number of particles per cell according to the ratio
+            // between their weight and the local weight integral
+            unsigned int          cell_index        = 0;
+            types::particle_index particles_created = 0;
+
+            for (const auto &cell : triangulation.active_cell_iterators())
+              if (cell->is_locally_owned())
+                {
+                  const types::particle_index particles_to_create =
+                    llround(static_cast<double>(n_local_particles) *
+                            accumulated_cell_weights[cell_index] /
+                            local_weight_integral);
+
+                  // We assume nobody creates more than 4 billion particles in
+                  // one cell
+                  particles_per_cell[cell_index] = static_cast<unsigned int>(
+                    particles_to_create - particles_created);
+                  particles_created += particles_per_cell[cell_index];
+                  ++cell_index;
+                }
+          }
+      }
+
+      // Now generate as many particles per cell as determined above
+      {
+        unsigned int cell_index             = 0;
+        unsigned int current_particle_index = start_particle_id;
+
+        // We first store the generated particles in a vector. Since they are
+        // generated cell-by-cell, they will already be sorted in the correct
+        // order to be later transferred to the multimap with O(N) complexity.
+        // If we would insert them into the multimap one-by-one it would
+        // increase the complexity to O(N log(N)).
+        std::vector<std::pair<Particles::internal::LevelInd, Particle<dim>>>
+          local_particles;
+        local_particles.reserve(n_local_particles);
+
+        std::multimap<
+          typename Triangulation<dim, spacedim>::active_cell_iterator,
+          Particle<dim, spacedim>>
+          particles;
+
+        for (const auto &cell : triangulation.active_cell_iterators())
+          if (cell->is_locally_owned())
+            {
+              for (unsigned int i = 0; i < particles_per_cell[cell_index]; ++i)
+                {
+                  const Particle<dim, spacedim> particle =
+                    random_particle_in_cell(cell,
+                                            current_particle_index,
+                                            random_number_generator,
+                                            mapping);
+                  particles.insert(particles.end(),
+                                   std::make_pair(cell, particle));
```

Comment:
I think you can probably use `particles.emplace_hint (particles.end(), {cell, std::move(particle)});` here.